### ISSUE TITLE
Avionics discussions

### ISSUE THREAD

**Issue by brightproject** [makerplane/FIX-Gateway#134]
Good afternoon @birkelbach and @e100.
Many thanks to the authors of this project!
I would like to ask you to create a discussion section in the `GitHub` repository, and in it you can communicate, like on a forum, and immediately receive information regarding the project code, and this is more convenient than on the forum, and this is not an issue.
I wanted to ask about examples and details of using the file

> https://github.com/makerplane/FIX-Gateway/blob/master/fixgw/plugins/mgl/rdac.py

I wanted to display data from my `AHRS` and `RDAC` manufactured by `MGL`.
But due to the lack of sufficient experience and examples, I have difficulties.
I have two such modules:
![RDAC_xb-AHRS_sp3h](https://github.com/makerplane/FIX-Gateway/assets/1788098/74f976e1-8b3a-481d-9142-e29238736bee)
With AHRS, it is not yet clear what protocol comes out of it and whether the block works at all.
Regarding RDAK, as far as I understand, MGL used different protocols:
1. OEM protocol

> https://github.com/makerplane/FIX-Gateway/files/15090712/RDAC_X_OEM_protocol.pdf

2. Raw data format (essentially a similar protocol, but there are differences)

> https://github.com/makerplane/FIX-Gateway/files/15090688/MGL_EFIS_data_feed.pdf

Please tell me which protocol is implemented in the [code](https://github.com/makerplane/FIX-Gateway)?

**Comment by e100** [comment]
Pyefis gets all of the data from the fix gateway using net-FIX protocol  https://github.com/makerplane/FIX-Gateway/blob/master/doc/plugins/netfix.rst
Net-FIX is just the TCP implementation of the CAN-FIX protocol https://github.com/makerplane/canfix-spec

The FIX-gateway has various plug-ins to communicate with different devices to send or receive data.

Regarding the MGL RDAC I have created a plug-in for the MGL CAN protocol https://www.mglavionics.co.za/Docs/MGL%20CAN%20Protocol.pdf

Only the RDAC XF portion of the CAN protocol is implemented, this starts on PDF page 11 in the above document. I own an RDAC XG and have tested the plug-in using it, it uses same protocol as thr RDAC XF. The plugin can be configured to receive data from the RDAC or to send data as an emulated RDAC.

I am unfamiliar with the RDAC XB that you have and based on the documentation you referenced it does not seem to use CAN for communication so is not supported by the plugin. I did see the RDAX XB listed on MGLs legacy product page https://www.mglavionics.co.za/legacy2.html

I can see two options if you want to interface with the legacy RDAC XB:
1. Write a plug-in yourself, or find someone who can write it. I currently do not have time to help with writing such a plug-in since I have enough work building my airplane and efis to go in it.
2. Update your hardware to an RDAC XG or RDAC XF. Upgrading might also require updating other MGL components too, so this might not be a practical option.

Regarding the MGL SP-3, it also seems to be a legacy product. It is not listed under their current products or their legacy products but they did have SP-2 and SP-4 listed as legacy.

A newer AHRS sensor that supports the MGL CAN protocol is not currently supported but I think the RDAC plug-in could be modified to support other MGL CAN devices like the SP-6 and SP-7.  I named the plugin 'mgl'  expecting that it would be expanded in the future to support more MGL CAN devices.
To implement a table would need created for them: https://github.com/makerplane/FIX-Gateway/blob/master/fixgw/plugins/mgl/tables.py
And the plugin would need updated to support the device/table.

Example to configure the MGL RDAC plug-in can be found in the default configuration file:
https://github.com/makerplane/FIX-Gateway/blob/96318e2cee1e9df0055a5c2c7eff6b6f9558f63c/fixgw/config/default.yaml#L404

**Comment by e100** [comment]
Also,  we have a forum:
https://www.makerplane.org/forum/

**Comment by brightproject** [comment]
> I can see two options if you want to interface with the legacy RDAC XB

Yes, I am on the forum, but there are mainly questions of a general nature and from pilots or engineers who install/repair devices.
There is still a community of programmers on `Github`, and I am more interested in code issues.
I will try to write the plugin code for outdated/legacy `RDAC` myself, I just need to understand how I can send the protocol data from the device via `USB-TTL` to `USB` and then to `pyEfis`.
`RDAC` genering a protocol in `RS-232` format, I receive it using a `USB` adapter.
![Protocol](https://github.com/makerplane/FIX-Gateway/assets/1788098/9012a75d-6e16-4558-b666-296d3bc10d48)

When the `FIX-Gateway` starts, it displays the following data about the connection port:
```
test1@test1-VirtualBox:~/FIX-Gateway$ ./fixgw.py -v
INFO:20240424-14:31:23:fixgw - Starting FIX Gateway
INFO:20240424-14:31:23:fixgw - Configuration Found at /home/test1/makerplane/fixgw/config/default.yaml
INFO:20240424-14:31:24:fixgw - Setting Initial Values - /home/test1/makerplane/fixgw/config/default.ini
None
INFO:20240424-14:31:24:fixgw.netfix - Initializing
INFO:20240424-14:31:24:fixgw.system - Initializing
INFO:20240424-14:31:24:fixgw.annunciate - Initializing
INFO:20240424-14:31:24:fixgw.compute - Initializing
INFO:20240424-14:31:24:fixgw.demo - Initializing
INFO:20240424-14:31:24:fixgw.netfix - Starting
INFO:20240424-14:31:24:fixgw.system - Starting
INFO:20240424-14:31:24:fixgw.annunciate - Starting
INFO:20240424-14:31:24:fixgw.compute - Starting
INFO:20240424-14:31:24:fixgw.demo - Starting
INFO:20240424-14:31:24:fixgw.netfix - Client connection from 127.0.0.1 port 56274
^CINFO:20240424-14:32:44:fixgw.netfix - Disconnected by 127.0.0.1 port 56274
INFO:20240424-14:32:44:fixgw.netfix - Stopping
INFO:20240424-14:32:45:fixgw.system - Stopping
INFO:20240424-14:32:45:fixgw.annunciate - Stopping
INFO:20240424-14:32:45:fixgw.compute - Stopping
INFO:20240424-14:32:45:fixgw.demo - Stopping
INFO:20240424-14:32:45:fixgw - FIX Gateway Exiting Normally
```
But how can I get data to `pyEfis` from the `com` port?
This is one of the things that is unclear to me now...

**Comment by e100** [comment]
Edit the configuration file, the path is shown on the output you provided.
Change demo load option to no so it does not load.

To get the data from you RDAC over serial you will need to create the plugin and then edit the config file to load it.

This plugin gets serial data from a Grand Rapids EIS, might be helpful to look at to understand how to create a plugin that talks to serial port:
https://github.com/makerplane/FIX-Gateway/blob/master/fixgw/plugins/grand_rapids_eis/__init__.py

This other plugin also uses serial port but not directly so not too useful as a serial port example. However this try catch in the main loop that initiates the serial connection allows the plugin to gracefully reconnect to the port in situations where the USB serial adapter is unplugged and then reconnected.
The Grand Rapids plugin does not have such logic and looks like it would likely crash the plugin in such a situation. 
https://github.com/makerplane/FIX-Gateway/blob/96318e2cee1e9df0055a5c2c7eff6b6f9558f63c/fixgw/plugins/mavlink/__init__.py#L59
The logic in the mavlink will also handle starting fix gateway without the USB connected, then automatically start collecting data when it is plugged in.

This is not required for the plugin and I'm no expert on udev so would not be much help but if you will have more than one USB serial port for different devices I highly recommend creating udev rules to give each port a static unique name. Without this you could find that one day your RDAC is ttyUSB0 and some other day it becomes ttyUSB1. My suggestion would be to name them ttyUSB10, ttyUSB11, ttyUSB12, etc for two reasons 1. You cannot rename ttyUSB1 to ttyUSB0 if ttyUSB0 already exists, so renaming to known unused numbers is more reliable and 2. Some security settings, like in apparmor or selinux, will sometimes let you connect to ttyUSB10,  etc. but not myCustomMadeUpDeviceName so using standard naming convention is more reliable. 
I had to write a similar udev rule for my two CAN0 and CAN1 bus ports because they kept flipping names randomly. I named them CAN10 and CAN11
https://github.com/e100/makerplane-configs/blob/main/INSTALL.md#optional---setup-waveshare-can-fd-hat

**Comment by brightproject** [comment]
> Edit the configuration file, the path is shown on the output you provided.

Thank you for your work and help, I will try to implement the plugin.
